dmenu-rs: add package

benjaminedwardwebb opened the pull request:
dmenu-rs is a pixel perfect port of dmenu, rewritten in Rust with extensive plugin support.

See: https://github.com/Shizcow/dmenu-rs

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

benjaminedwardwebb commented:
Ran `nix-env -iA dmenu-rs -f .` on the `nixos-unstable` branch at `52b2ac8ae18bbad4374ff0dd5aeee0fdf1aea739` in my local `nixpkgs` repository on my `aarch64-linux` laptop. Installed successfully and was usable.

benjaminedwardwebb commented:
Also ran the above with `nix.useSandbox = true;`.

benjaminedwardwebb commented:
Also, anecdotally, I've been using this (or [a largely similar derivation][1]) as my daily driver launcher for about a week and have observed no issues.

The overlay I use is [here][2].

[2]: https://codeberg.org/benjaminedwardwebb/backpack/src/branch/main/pkgs/dmenu-rs.nix
[1]: https://github.com/benjaminedwardwebb/nixpkgs/tree/dmenu-rs-alternative/pkgs/applications/misc/dmenu-rs

benjaminedwardwebb commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,111 @@
+    make install
+  '';
+
+  # Tests require an X11 server and yields a cannot open display error. It's
+  # probably possible to get an X server running in the build but seems hard.
+  doCheck = false;
+  checkPhase = ''
+    make test
+  '';
```
While I'd like to enable this, the test triggered by `make test` invokes `dmenu` directly.

[It runs][1]

```bash
seq 1 100 | target/dmenu $(ARGS)
```

This not only requires an X11 server but it also hangs waiting for user input.

I was able to hack together the following `checkPhase` that works locally using `timeout`

```bash
timeout --preserve-status 5s xvfb-run make test
```

however simply hoping that 5 seconds is long enough for xvfb-run to spin up and the test to complete on every machine seems like an awful idea. I'm not sure how to kill the hanging call to `dmenu`, no option seems to allow that and I don't know how to feed in a keypress like `<RETURN>` into the `xvfb-run` context.

So I'm just leaving it disabled and relying on the other tests like installing, running the binary, and using it personally on a regular basis.

benjaminedwardwebb reviewed (commented):
@ElvishJerricco @symphorien thanks for your help in working through some issues that stumped me while working on this derivation.

Of course not a priority, but if either of you have write-access and think the changes are worth merging, I'd appreciate getting this package out of a personal overlay and into `nixpkgs` where others can use it.

ElvishJerricco commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,111 @@
+  '';
+
+  RUST_BACKTRACE = "full";
+  LIBCLANG_PATH = "${libclang.lib}/lib";
```
Why are these needed?

symphorien replied:
to provide clang and the right files to bindgen, use rustPlaftorm.bindgenHook instead.

benjaminedwardwebb replied:
That worked! Thanks for the tip about `rustPlatform.bindgenHook`.

ElvishJerricco commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,111 @@
+  # and make it writeable.
+  # See: https://github.com/rust-x-bindings/rust-xcb/tree/v0.8.2
+  patchPhase = ''
+    chmod +w /build/cargo-vendor-dir
```
Do not use `/build`. It is different depending on the machine. Use `$NIX_BUILD_TOP`.

ElvishJerricco commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,111 @@
+    sed -i "s@PREFIX = /usr/local@PREFIX = $out@g" config.mk
+  '';
+
+  buildPhase = ''
+    make
+  '';
+
+  installPhase = ''
+    make install
+  '';
```
I thought these looked fishy at first, but it actually does make sense. We want the rust platform stuff because it does the vendoring and all that, but it doesn't just do `make` and `make install` like `stdenv.mkDerivation` does by default.

That said, this is very much *not* going to cross compile correctly. Not critical, but unfortunate, and probably fixable.

benjaminedwardwebb replied:
I've added a comment that tries to explain these phases' surprising appearance.

I'll be honest, I don't know the exact consequences of this package not cross compiling correctly, but it sure sounds bad. Is it related to not running `./configure` in the `configurePhase`? Should I go read up on Autoconf and figure out how to add that into this derivation?

ElvishJerricco replied:
I wouldn't worry about it not cross compiling; frankly I don't think you can cross compile a desktop environment with nixpkgs right now anyway. It was just something I found noteworthy. But basically the package's makefile is hostile to cross compilation because e.g. it hardcodes `CC=cc` and doesn't have a way to choose the cargo target. It would need (relatively trivial) changes to their build scripting to fix this.

benjaminedwardwebb replied:
Thanks for calljng it out. I'm not really a contributor to `dmenu-rs` (I fixed 1 typo) but it's something I'll keep in mind and who knows maybe I can fix it upstream some time in the future.

symphorien replied:
did you consider using mkDerivation with cargoSetupHook to handle fetching sources?

to quote the manual:
> The `rustPlatform.buildRustPackage` function is split into several hooks: cargoSetupHook to set up vendoring for Cargo-based projects, cargoBuildHook to build a project using Cargo, cargoInstallHook to install a project using Cargo, and cargoCheckHook to run tests in Cargo-based projects. With this change, mixed-language projects can use the relevant hooks within builders other than `buildRustPackage`. However, these changes also required several API changes to `buildRustPackage` itself:

benjaminedwardwebb replied:
I didn't. That sounds like a more natural way to write this derivation -- as a standard make-based build that happens to include cargo deps. I'll take a stab at rewriting it that way later today. Thanks for this!

ElvishJerricco reviewed (approved):
I'm a little unhappy with the `LIBCLANG_PATH` thing; I don't really understand why this package is so picky about using clang and having this lib in this particular environment variable. But it's probably not worth blocking this PR over given that it's probably due to the package's excessively wonky build system.

symphorien commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,113 @@
+    cp -r "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2/." "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite"
+    unlink "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2"
+    mv "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite" "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2"
+    chmod -R +w "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2/"
```
why do you need the copy step, and not only the recursive chmod?

benjaminedwardwebb replied:
`.../cargo-vendor-dir/xcb-0.8.2/` is a symlink to the nix store. I don't think I can `chmod` the nix store inside of the build and in general making anything in the nix store writeable seems conceptually wrong?

symphorien replied:
oh it's originally a symlink, I thought it was a copy. My bad.

you can remove the final chmod -R by using cp --no-preserve=mod then

symphorien commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,113 @@
+  };
+
+  nativeBuildInputs = [
+    clang
```
with bindgenHook you presumably need neither libclang, llvm nor clang.

benjaminedwardwebb replied:
I was able to drop `llvm` and `libclang`, but it wouldn't build without keeping `clang`.

symphorien replied:
please drop clang and modify https://github.com/Shizcow/dmenu-rs/blob/master/config.mk#L14 as suggested by the error message

benjaminedwardwebb replied:
Are you suggesting I modify this upstream at `Shizcow/dmenu-rs` or through a patch in the nix derivation?

symphorien replied:
downstream is ok (either as a patch or as `substituteInplace config.mk --replace 'clang' '$(CC)'` in postPatch for example)
But hard coding clang as compiler is not distro-friendly in general so you may submit this change upstream if so you wish, after we merge this PR. Using $(CC) is more portable and preferable.

benjaminedwardwebb replied:
I've added it as a patch here, but also opened [this PR upstream](https://github.com/Shizcow/dmenu-rs/pull/44).

symphorien replied:
I think you forgot to remove the dependency on clang. Thank you for reaching upstream :)

benjaminedwardwebb replied:
Oh you're right, woops

benjaminedwardwebb replied:
Removed

symphorien commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,113 @@
+  # We use buildRustPackage instead of mkDerivation to vendor the cargo
+  # dependencies. However dmenu-rs is not built with cargo, it's built using a
+  # standard makefile.
+  buildPhase = ''
```
when you override a full phase, please start with `runHook preBuild` and end with `runHook postBuild`. Same for installPhase.

symphorien commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,84 @@
+  '';
+
+  # Running make test requires an X11 server, so we skip it.
+  doCheck = false;
```
you can probably run tests with xvfb-run

benjaminedwardwebb replied:
I've updated the `checkPhase` to use the following

```bash
timeout 30s xvfb-run make test || if [ "$?" == "124" ]; then true; else false; fi
```

but ... personally, I would prefer no check phase at all.

This test only passes if the 30 seconds is long enough for the test to run. It seemed to work pretty consistently with 5 seconds on my machine, but using a timeout like this seems destined to cause flaky failures somewhere.

The reason I reached for a timeout was because the `make test` hangs on user input, and I don't how to supply any input keys to the program or avoid this hanging test.

For more detail see: https://github.com/NixOS/nixpkgs/pull/200977#discussion_r1027367488

symphorien replied:
My bad I had no read the full discussion. If the tests are not well behaved enough to terminate, let's forget about them.

benjaminedwardwebb replied:
Cool, dropped the test.

AndersonTorres commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,103 @@
+# only builds the package with the default set of plugins. If you'd like to
+# further customize dmenu-rs you can build it from the source.
+# See: https://github.com/Shizcow/dmenu-rs#plugins
+stdenv.mkDerivation rec {
```
Boy, this is very confusing!

benjaminedwardwebb replied:
Definitely. I started using `buildRustPackage` initially, until @symphorien suggested I use `mkDerivation` and it made the build a lot simpler.

AndersonTorres commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,103 @@
+    runHook postPatch
+  '';
+
+  # As suggested in the nixpkgs manual section on building rust packages, the
+  # derivation's Cargo.lock file is copied into the source.
```
```suggestion
```

Mentioning the manual is not needed.

AndersonTorres commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,103 @@
+    cp ${./Cargo.lock} src/Cargo.lock
+  '';
+
+  # The cargo project does not live at the repository's top-level directory.
```
Too obvious to be commented.

```suggestion
```

benjaminedwardwebb replied:
Thanks for keeping things concise! :)

benjaminedwardwebb committed:
make me a nixpkgs maintainer

benjaminedwardwebb committed:
dmenu-rs: init at 5.5.1

dmenu-rs is a pixel perfect port of dmenu, rewritten in Rust with
extensive plugin support.

See: https://github.com/Shizcow/dmenu-rs

benjaminedwardwebb commented on maintainers/maintainer-list.nix:
```diff
@@ -1671,6 +1671,15 @@
       fingerprint = "D35E C9CE E631 638F F1D8  B401 6F0E 410D C3EE D02";
     }];
   };
+  benjaminedwardwebb = {
```
Done :)

AndersonTorres commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,98 @@
+
+  cargoRoot = "src";
+
+  preConfigure = ''
+    sed -i "s@PREFIX = /usr/local@PREFIX = $out@g" config.mk
+  '';
```
Some more corners to 

1. this should be a postPatch
2. It is better to remove this setting from config.mk:

```suggestion
  postPatch = ''
    sed -i "s@PREFIX = /usr/local@@g" config.mk
  '';
```
3. Use `installFlags = [ "PREFIX=$(out)" ];`.  (Indeed, maybe the previous step becomes unneeded if this flag works.)

AndersonTorres commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,98 @@
+
+  cargoRoot = "src";
+
+  preConfigure = ''
```
Some more corners to trim

1. this should be a postPatch
2. It is better to remove this setting from config.mk: `sed -i "s@PREFIX = /usr/local@@g" config.mk`
3. Way better: remove this preConfigure completely, and use `installFlags = [ "PREFIX=$(out)" ];` instead.

benjaminedwardwebb replied:
I actually was trying to go with something like 3 initially, but then decided to just copy what the vanilla C-based `dmenu` was doing in its build: https://github.com/NixOS/nixpkgs/blob/d0a1a0f243138978b1d258056eb5988787da7aac/pkgs/applications/misc/dmenu/default.nix#L21-L23

But I agree, `installFlags` seems more sensible. Let me see if I can get it working.

benjaminedwardwebb replied:
Cool, updated to use `installFlags`. I think the first time I was trying to use `makeFlags` and couldn't figure out why it wasn't working -- thanks for suggesting the right way.

symphorien commented on pkgs/applications/misc/dmenu-rs/default.nix:
```diff
@@ -0,0 +1,96 @@
+    runHook postPatch
+  '';
+
+  postPatch = ''
```
there is no reason to have both postPatch and a patchPhase. As the default patchPhase is still useful if someone downstream wants to add a patch with overrideAttrs, it is probably better to move the content of patchPhase (except runHook) to postPatch.

benjaminedwardwebb replied:
OK. I've moved everything over into `postPatch`.

But, any guidance on what is a good definition of a patch is / what should generally go in `patchPhase`?

Should it only be applying `.patch` files? Or any change to the project's source code? Any change to the project's build? Any change to the project's dependencies' source code or build?

I guess what this `patchPhase` is currently doing is
  - changing the project's build code (`config.mk` and copying in `Cargo.lock`)
  - subtly change a vendored dependency so that it can be written to, without actually altering any of that dependency's source code or build code

So if drew the line at "patches are things that alter a project's source code (but not necessarily build code)" then it seems like it makes sense to keep it all out of `patchPhase`.

Probably there's no hard rule but just trying to get a feel for what is what in nixpkgs.

AndersonTorres replied:
> But, any guidance on what is a good definition of a patch is / what should generally go in `patchPhase`?
> 
> Should it only be applying `.patch` files? Or any change to the project's source code? Any change to the project's build? Any change to the project's dependencies' source code or build?

Typically, many distros (if not all!) provide `.patch` files in order to directly modify the source code of the apps they provide. 

A typical case is when a new compiler release does not compile some old code (indeed it happened to me when I tried to compile qemu in Slackware a long time ago...). So, usually a package manager provides patch files for cases like these.
Indeed, many times those patchsets are generated by merely diffing the `master` git branch with the current `stable` git branch, maybe because some people like ultra fresh, pre-alpha code...

Another thing that such package managers can do too is run sanitization scripts that clean up some quirks on the original source code; things like removing annoying lines from a Makefile or fixing hardcoded locations (e.g. FreeBSD ports system installs things on `/usr/local/` instead of `/usr/`; so maybe some tool that expects to find things on /usr will not get it without help).

It gets a bit complicated when both things need to be done: `.patch` files to modify C source code, plus sanitization scripts that cleanup Makefiles.

Now, switch to Nix(pkgs).

In Nixpkgs, we provide both use cases via the same `patchPhase`. However, as I said above, it is a bit complicated...

The typical, unedited patchPhase usually works like this:

```nix
patchPhase = ''
  runHook prePatch

  patch -p1 < <the .patch files provided in patches =[ ... ];>
 
  runHook postPatch
''
```

The `runHook`s are there because of composability:

Nix has the ability of compose overlays of package trees, in a similar manner when you add third party repositories in other package managers (`/etc/apt/sources.lst` anyone?).
Such pre and post hooks allow the creator of an overlay to easily modify the phases without having to rewrite them completely.

Well, when you customize this phase as above, it would look like

```nix
patchPhase = ''
  runHook prePatch

  <a giant script modifying the source code>
 
  runHook postPatch
''
```

However, by doing this, you are basically ignoring any `.patch` files eventually provided in `patches` list.

So, when you need to use such "sanitization scripts", it is better to use `postPatch` instead of rewriting the whole `patchPhase`.

> Probably there's no hard rule but just trying to get a feel for what is what in nixpkgs.

Sometimes many things are developed by trial and error.

benjaminedwardwebb replied:
Ah thank you for all that good background knowledge!

symphorien commented:
Looks good now! Thank you for bearing with us. One last thing before I merge: this does not build on aarch64 (see https://github.com/NixOS/nixpkgs/pull/200977/checks?check_run_id=10128726771 ). It looks due to upstream but unintended, so please mark it as broken like here

https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/audio/eq10q/default.nix#L36

You can optionally report it upstream, and if they answer that aarch64 support is not in scope for the project, remove the broken attribute and restrict platforms to x86-64-linux only instead.

benjaminedwardwebb commented:
> Looks good now! Thank you for bearing with us. One last thing before I merge: this does not build on aarch64 (see https://github.com/NixOS/nixpkgs/pull/200977/checks?check_run_id=10128726771 ). It looks due to upstream but unintended, so please mark it as broken like here
> 
> https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/audio/eq10q/default.nix#L36
> 
> You can optionally report it upstream, and if they answer that aarch64 support is not in scope for the project, remove the broken attribute and restrict platforms to x86-64-linux only instead.

@symphorien added `aarch64-linux` as broken. However, I use an `aarch64-linux` machine pretty regularly, so am interested in adding support upstream if possible (and filed [this issue][1] related to it). I don't know how easy this will be, though, so maybe it will have to wait for another PR to nixpkgs.

[1]: https://github.com/Shizcow/dmenu-rs/issues/45

@AndersonTorres Anything holding up an approval?

benjaminedwardwebb commented:
Thanks all for the help reviewing and getting this merged! Adding my first package to `nixpkgs` was a very pleasant experience.
